tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php

```diff
@@ -865,86 +875,54 @@ private function skipOci8AndMysqli()
-            ->method('query')
-            ->will($this->returnCallback(function () use ($stmtSpy) {
-                $stmtSpy->execute();
-                return $stmtSpy;
-            }));
-
-        $stmt = $connectionMock->query($sql);
-        $stmt->setFetchMode(\PDO::FETCH_ASSOC);
-
-        // Assert no fetch calls thus far
-        self::assertEquals(0, $stmtSpy->getFetchCalls());
-
-        $i = 0;
-        foreach ($stmt as $row) {
-            $i++;
-        }
-
-        // Assert expected number of fetches
-        $this->assertGreaterThan(0, $stmtSpy->getFetchCalls());
-        $this->assertEquals($i, $stmtSpy->getFetchCalls());
+        return [
+            [MysqliConnection::class, MysqliStatement::class],
+            [OCI8Connection::class, OCI8Statement::class],
+            [DB2Connection::class, DB2Statement::class],
+            [SQLAnywhereConnection::class, SQLAnywhereStatement::class],
+            [SQLSrvConnection::class, SQLSrvStatement::class],
+        ];
     }
-}
 
-class StatementSpy implements \IteratorAggregate
-{
     /**
-     * @var Statement|\Doctrine\DBAL\Statement
+     * @dataProvider getConnections
      */
-    private $stmt;
+    public function testGettingIteratorDoesNotCallFetch($connection, $statement)
+    {
+        $stmt = $this->createPartialMock($statement, ['fetch']);
+        $stmt->expects($this->never())->method('fetch');
 
-    private $fetchCalls = 0;
+        $conn = $this->createMock($connection);
+        $conn->expects($this->any())->method('prepare')->willReturn($stmt);
 
-    public function __construct(Statement $stmt)
-    {
-        $this->stmt = $stmt;
+        $ii = new IteratorIterator($stmt);
+        $ii->getInnerIterator();
     }
 
-    public function setFetchMode($fetchMode, $arg2 = null, $arg3 = null)
+    /**
+     * @dataProvider getConnections
+     */
+    public function testIterationCallsFetchOncePerStep($connection, $statement)
     {
-        return $this->stmt->setFetchMode($fetchMode, $arg2, $arg3);
-    }
+        $values = ['foo', '', 'bar', '0', 'baz', 0, 'qux', null, 'quz', false];
+        $calls = 0;
 
-    public function execute($params = null)
-    {
-        return $this->stmt->execute($params);
-    }
+        $stmt = $this->createPartialMock($statement, ['fetch']);
+        $stmt->expects($this->exactly(10))
+            ->method('fetch')
+            ->willReturnCallback(function() use ($values, &$calls) {
+                $calls++;
+                yield from $values;
+            });
 
-    public function getIterator()
-    {
-        foreach ($this->stmt as $item) {
-            $this->fetchCalls++;
-            yield $item;
-        }
-    }
+        $conn = $this->createMock($connection);
+        $conn->expects($this->any())->method('prepare')->willReturn($stmt);
 
-    public function getFetchCalls()
-    {
-        return $this->fetchCalls;
+        foreach ($stmt as $i => $_) {
+            $_->next();
```

Comment:
This is not something I'd expect the code doing. Why should `next()` be called on the value returned by iterator? If I understand it correctly, `$stmt` is the iterator, and `foreach` will call `next()` on it.

Reply:
Yeah agreed, it didn't seem to iterate over without calling next though. Hopefully the changes you mention below about extracting the iterator will help here.